composer/datasets/streaming/download.py

```diff
@@ -0,0 +1,132 @@
+import os
+import shutil
+from time import sleep, time
+from urllib.parse import urlparse
+
+
+def wait_for_download(local: str, timeout: float = 20) -> None:
+    """Block until another worker's shard download completes.
+
+    Args:
+        local (str): Path to file.
+        timeout (float): How long to wait before raising an exception. Default: 20 sec.
+    """
+    start_time = time()
+    i = 0
+    while True:
+        if os.path.exists(local):
+            return
+        elapsed = time() - start_time
+        if elapsed > timeout:
+            raise TimeoutError(f'Waited too long (more than {timeout:.3f} sec) for download')
+        sleep(0.1)
+        i += 1
+
+
+def download_from_s3(remote: str, local: str, timeout: float) -> None:
+    """Download a file from remote to local.
+
+    Args:
+        remote (str): Remote path (S3).
+        local (str): Local path (local filesystem).
+        timeout (float): How long to wait for shard to download before raising an exception.
+    """
+    import boto3
+    from botocore import Config
+    obj = urlparse(remote)
+    if obj.scheme != 's3':
+        raise ValueError(f"Expected obj.scheme to be 's3', got {obj.scheme} for remote={remote}")
+
+    # We don't know how much of total 'timeout' to assign to connect vs. read
+    # So we allow both connect and read to take up to 'timeout' seconds
+    # And if the overall time is greater than 'timeout', our parent `download` function will catch it.
+    config = Config(connect_timeout=timeout, read_timeout=timeout, retries={'total_max_attempts': 5})
+    s3 = boto3.client('s3', config=config)
+    s3.download_file(obj.netloc, obj.path[1:], local)
+
+
+def download_from_local(remote: str, local: str) -> None:
+    """Download a file from remote to local.
+
+    Args:
+        remote (str): Remote path (local filesystem).
+        local (str): Local path (local filesystem).
+    """
+    shutil.copy(remote, local)
```

Comment:
I think we actually might want to copy... the situation I'm thinking of is remote is an NFS mount, and local is localdisk, in which case you do want to copy so that on epoch 2, you are reading from localdisk. Does that make sense?